### PR TITLE
Add GNU PE64 binary parser for MinGW GCC toolchain

### PR DESCRIPTION
**Pull request by jld01**
Part of #361

### CODE REVIEW

**Comment by jld01** [comment]
This PR introduces a new GNU PE64 parser for use with the MinGW GCC toolchain. It allows the user to double click on a MinGW executable and invoke `objdump` if present on the global `PATH`.

File: core/org.eclipse.cdt.core/plugin.xml [hunk 1]
```diff
@@ -72,6 +72,16 @@
          </run>
       </cextension>
    </extension>
+   <extension
+         id="GNU_PE64"
```
Comment (Torbjorn-Svensson) [inline core/org.eclipse.cdt.core/plugin.xml:76]: With the addition of this new parser, I suppose the [documentation](https://github.com/eclipse-cdt/cdt/blob/main/build/org.eclipse.cdt.managedbuilder.core/schema/ManagedBuildTools.exsd#L517) needs to be updated accordingly.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	public GNUPEBinaryObject64(IBinaryParser parser, IPath path, ARHeader header) {
+		super(parser, path, header);
+	}
+
+	public GNUPEBinaryObject64(IBinaryParser parser, IPath path, int type) {
+		super(parser, path, type);
+	}
+
+	@Override
+	public InputStream getContents() throws IOException {
```
Comment (Torbjorn-Svensson) [inline core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java:40]: I would suggest to use the following content of this function instead as it, in my opinion, makes it a bit easier to see what the return value will be.

```java
		Objdump objdump = getObjdump();
		if (objdump != null) {
			try {
				byte[] contents = objdump.getOutput();
				return new ByteArrayInputStream(contents);
			} catch (IOException e) {
				// Nothing
			}
		}
		return super.getContents();
```

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	protected IBinaryObject createBinaryObject(IPath path) {
+		return new GNUPEBinaryObject64(this, path, IBinaryFile.OBJECT);
+	}
+
+	@Override
+	protected IBinaryShared createBinaryShared(IPath path) {
+		return new GNUPEBinaryShared64(this, path);
+	}
+
+	protected DefaultGnuToolFactory createToolFactory() {
```
Comment (Torbjorn-Svensson) [inline core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:59]: Should this method have return type `DefaultGnuToolFactory` or `IGnuToolFactory`?

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.IOException;
+
+import org.eclipse.cdt.utils.DefaultGnuToolFactory;
+import org.eclipse.cdt.utils.IGnuToolFactory;
+import org.eclipse.core.runtime.IPath;
+
+/** @since 8.2 */
+public class GNUPEParser64 extends PEParser64 {
+
+	private DefaultGnuToolFactory toolFactory;
```
Comment (Torbjorn-Svensson) [inline core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:27]: Should this field be of type `DefaultGnuToolFactory` or `IGnuToolFactory`?

File: build/org.eclipse.cdt.managedbuilder.gnu.ui/plugin.xml [hunk 5]
```diff
@@ -2388,7 +2388,7 @@
                   <targetPlatform
 					  id="cdt.managedbuild.target.gnu.platform.mingw.base"
 				 	  name="%PlatformName.Dbg"
-                      binaryParser="org.eclipse.cdt.core.PE64"            					  
+                      binaryParser="org.eclipse.cdt.core.GNU_PE64"
```
[the remark this answers is not in the document]
Comment (jld01) [reply build/org.eclipse.cdt.managedbuilder.gnu.ui/plugin.xml:2391]: @Torbjorn-Svensson the entire file needs whitespace clean up. Suggest we perform this as a separate commit. Does that work for you?

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java (hunk 2 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java:40]: @Torbjorn-Svensson this is just a copy/paste from another implementation. Is clarity or consistency more important here?

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java (hunk 3 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:59]: `IGnuToolFactory` would be adequate, but inconsistent with other implementations. Do you think it worth changing them all?

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java (hunk 4 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:27]: IGnuToolFactory would be adequate, but inconsistent with other implementations. Do you think it worth changing them all?

Comment (Torbjorn-Svensson) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:27]: What other implementations are you referring to?
The [GNUElfParser](https://github.com/eclipse-cdt/cdt/blob/main/core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/elf/parser/GNUElfParser.java) uses the interface as an example.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java (hunk 3 above)
Comment (Torbjorn-Svensson) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:59]: What other implementations are you referring to?
The [GNUElfParser](https://github.com/eclipse-cdt/cdt/blob/main/core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/elf/parser/GNUElfParser.java) uses the interface as an example.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java (hunk 2 above)
Comment (Torbjorn-Svensson) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java:40]: Aha, okay. Well in that case, either would do.
I still think my suggestion is clearer, but I suppose consistency is good to have too...
How many other places is the same snippet available at?

File: core/org.eclipse.cdt.core/plugin.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jld01) [reply core/org.eclipse.cdt.core/plugin.xml:76]: Yes, I will update the doc.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java (hunk 2 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEBinaryObject64.java:40]: I copied from `GNUElfBinaryObject`. That may be the only class with that exact implementation.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java (hunk 3 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:59]: Yes. I copied and modified from `CygwinPEParser64`. I will change this as you suggest.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java (hunk 4 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java:27]: I will change this.

File: core/org.eclipse.cdt.core/plugin.xml (hunk 1 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/plugin.xml:76]: Duplicate comment.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java (hunk 4 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java]: Done.

File: core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java (hunk 3 above)
Comment (jld01) [reply core/org.eclipse.cdt.core/utils/org/eclipse/cdt/utils/coff/parser/GNUPEParser64.java]: Done.

File: core/org.eclipse.cdt.core/plugin.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jld01) [reply core/org.eclipse.cdt.core/plugin.xml:76]: Done.

**Review by Torbjorn-Svensson — APPROVED** [review]
I have not run the code, but based on visual review, I think this is fine.
Thank you John for working on this!

**Comment by jld01** [comment]
> I have not run the code, but based on visual review, I think this is fine. Thank you John for working on this!

Thank you for the review, @Torbjorn-Svensson. I will merge this and move on to per-config environment setup.

**Comment by baoruiqi** [comment]
@jld01 Thanks for your effort, this is a very useful feature! Based on this, I have a question.

I downloaded the latest Embedded IDE from Eclipse's website, created a C managed project and selected the riscv-cross toolchain. In the project properties, I set the local toolchain path, and the binary file of the risc-v architecture can be successfully generated after the build.

However, when I double-click the elf file, the editor view only shows the header information, and there is no disassembly content.

image

Usually the disassembly is obtained through the objdump tool related to the toolchain, passing the -d parameter. However, the binary parser tab in the project configuration interface provided by CDT does not provide relevant parameter configuration entries.

If I want to double-click the binary file and display the disassembly, do I need to modify the cdt?

Below is my IDE info:
```
Eclipse IDE for Embedded C/C++ Developers (includes Incubating components)
Version: 2023-06 M2 (4.28.0 M2)
Build id: 20230504-1428
OS: Linux, v.5.19.0-40-generic, x86_64 / gtk 3.24.33, WebKit 2.38.6
Java vendor: Eclipse Adoptium
Java runtime version: 17.0.7+7
Java version: 17.0.7
```

**Pull request merged**